lib/linter.js

```diff
@@ -814,27 +814,35 @@ class Linter extends EventEmitter {
 
             // if this result is from a parseForESLint() method, normalize
             if (parseResult && parseResult.ast) {
-                ast = parseResult.ast;
-            } else {
-                ast = parseResult;
-                parseResult = null;
-            }
 
-            if (ast) {
-                this.sourceCode = new SourceCode(text, ast);
+                // Save all returned values of `parse()` because this source code object can be reused.
+                // This ensures the same result when reused.
+                this.sourceCode = new SourceCode(
+                    text,
+                    parseResult.ast,
+                    parseResult.services,
+                    parseResult.scope,
```

Comment:
Before this PR is merged, we should add documentation for the new parser APIs. (However, maybe we should wait until we decide on the APIs before adding documentation.)